Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,553 @@
+                self.logger.log("Device Id could not be retrieved",
+                                level: .warning,
+                                category: .sync)
+                deferred.fill(Maybe(failure: DeviceIdError()))
+                return
+            }
+
+            accountManager.getAccessToken(scope: OAuthScope.oldSync) { result in
+                guard let accessTokenInfo = try? result.get(),
+                      let key = accessTokenInfo.key else {
+                    deferred.fill(Maybe(failure: ScopedKeyError()))
+                    return
+                }
+
+                accountManager.getTokenServerEndpointURL { result in
+                    guard case .success(let tokenServerEndpointURL) = result else {
+                        deferred.fill(Maybe(failure: NoTokenServerURLError()))
+                        return
+                    }
+
+                    self.getEnginesAndKeys(engines: engines).upon { result in
+                        guard let (rustEngines, localEncryptionKeys) = result
+                            .successValue else {
+                            deferred.fill(Maybe(failure: EngineAndKeyRetrievalError()))
+                            return
+                        }
+                        let params = SyncParams(
+                            reason: why,
+                            engines: SyncEngineSelection.some(engines: rustEngines),
+                            enabledChanges: self.getEngineEnablementChangesForAccount(),
+                            localEncryptionKeys: localEncryptionKeys,
+                            authInfo: SyncAuthInfo(
+                                kid: key.kid,
+                                fxaAccessToken: accessTokenInfo.token,
+                                syncKey: key.k,
+                                tokenserverUrl: tokenServerEndpointURL.absoluteString),
+                            persistedState:
+                                self.prefs
+                                    .stringForKey(PrefsKeys.RustSyncManagerPersistedState),
+                            deviceSettings: DeviceSettings(
+                                fxaDeviceId: device.id,
+                                name: device.displayName,
+                                kind: self.toSyncManagerDeviceType(
+                                    deviceType: device.deviceType)))
+
+                        self.beginSyncing()
+                        self.syncManagerAPI.sync(params: params) { syncResult in
+                            guard syncResult.status == ServiceStatus.ok else {
+                                deferred.fill(Maybe(success: syncResult))
+                                self.endSyncing(syncResult)
+                                return
+                            }
+
+                            // Save the persisted state
+                            self.prefs
+                                .setString(syncResult.persistedState,
```

Comment:
It's a bit of a bug in the rust code - `syncResult.persistedState` should probably be nullable to make the intent clearer, but I think you should persist this value any time it is not empty, even if the sync "failed"